File: CodingNinjas/Stacks/balancedParentheses.cpp
#include<bits/stdc++.h>
using namespace std;

bool isBalanced(string expression) {
    stack<char> s;
    int i=0;
    if(expression[i]==')'){return false;}
    while(i<expression.length()){
        if(expression[i]=='('){s.push(expression[i]);i++;}
        else if(expression[i]==')'){s.pop();i++;}
        else{i++;}
    }
    return s.empty();
}

int main(){
	freopen("input.txt","r",stdin);
	freopen("output.txt","w",stdout);

	string s;cin>>s;
	cout<<boolalpha<<isBalanced(s);
	return 0;
}

File: DSAPython/Stacks/balancedParentheses.py
def isBalanced(string):
    stack =[]
    for char in string:
        if char in "({[":
            stack.append(char)
        elif char == ')':
            if not stack or stack[-1] != '(':
                return False
            stack.pop()
        elif char == '}':
            if not stack or stack[-1] != '{':
                return False
            stack.pop()
        elif char == ']':
            if not stack or stack[-1] != '[':
                return False
            stack.pop()
    if not stack:
        return True
    return False
 
string = input("Enter the string: ")
ans = isBalanced(string)
print(ans)